File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/dispatcher/constants/KeyLocationMapper.java
package io.appium.espressoserver.lib.helpers.w3c.dispatcher.constants;

public class KeyLocationMapper {
    /**
     * Implement table found in 17.4.2 that maps characters to 'location'
     *
     * Not using constants or enums because I don't think we'll actually need these in any
     * specific implementation
     * @param key Raw key
     * @return The location of the key or default 0
     */
    public static int getLocation(String key) {
        switch (key) {
            case "\uE007": return 1;
            case "\uE008": return 1;
            case "\uE009": return 1;
            case "\uE00A": return 1;
            case "\uE01A": return 3;
            case "\uE01B": return 3;
            case "\uE01C": return 3;
            case "\uE01D": return 3;
            case "\uE01E": return 3;
            case "\uE01F": return 3;
            case "\uE020": return 3;
            case "\uE021": return 3;
            case "\uE022": return 3;
            case "\uE023": return 3;
            case "\uE024": return 3;
            case "\uE025": return 3;
            case "\uE026": return 3;
            case "\uE027": return 3;
            case "\uE028": return 3;
            case "\uE029": return 3;
            case "\uE03D": return 1;
            case "\uE050": return 2;
            case "\uE051": return 2;
            case "\uE052": return 2;
            case "\uE053": return 2;
            case "\uE054": return 3;
            case "\uE055": return 3;
            case "\uE056": return 3;
            case "\uE057": return 3;
            case "\uE058": return 3;
            case "\uE059": return 3;
            case "\uE05A": return 3;
            case "\uE05B": return 3;
            case "\uE05C": return 3;
            case "\uE05D": return 3;
            default: return 0;
        }
    }
}

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/dispatcher/constants/KeyLocationMapper.kt
package io.appium.espressoserver.lib.helpers.w3c.dispatcher.constants

/**
 * Implement table found in 17.4.2 that maps characters to 'location'
 *
 * Not using constants or enums because I don't think we'll actually need these in any
 * specific implementation
 * @param key Raw key
 * @return The location of the key or default 0
 */
fun getLocation(key: String?): Int {
    return when (key) {
        "\uE007" -> 1
        "\uE008" -> 1
        "\uE009" -> 1
        "\uE00A" -> 1
        "\uE01A" -> 3
        "\uE01B" -> 3
        "\uE01C" -> 3
        "\uE01D" -> 3
        "\uE01E" -> 3
        "\uE01F" -> 3
        "\uE020" -> 3
        "\uE021" -> 3
        "\uE022" -> 3
        "\uE023" -> 3
        "\uE024" -> 3
        "\uE025" -> 3
        "\uE026" -> 3
        "\uE027" -> 3
        "\uE028" -> 3
        "\uE029" -> 3
        "\uE03D" -> 1
        "\uE050" -> 2
        "\uE051" -> 2
        "\uE052" -> 2
        "\uE053" -> 2
        "\uE054" -> 3
        "\uE055" -> 3
        "\uE056" -> 3
        "\uE057" -> 3
        "\uE058" -> 3
        "\uE059" -> 3
        "\uE05A" -> 3
        "\uE05B" -> 3
        "\uE05C" -> 3
        "\uE05D" -> 3
        else -> 0
    }
}